### ISSUE TITLE
test_clear fails in while loop

### ISSUE THREAD

**Issue by dokipen3d** [openglonmetal/MGL#43]
This may or may not be a simple one. It seems that if we run the contents of test_clear function in a while loop, then we get this runtime error repeatedly.

`[CAMetalLayer nextDrawable] returning nil because allocation failed`

Not sure why, as other test programs can clear. They have extra stuff added though, so the process of adding programs and changing other state, must initialize metal correctly, whereas not doing those things makes clearing by itself not possible.

I'll keep digging and learn a few things long the way :)

**Comment by dokipen3d** [comment]
Seems that in MGLRenderer.m newRenderEncoder() we return if vao is NULL. removing that allows the layer to be allocated correctly. Next issue that shows up is that 
```Assertion failed: (ptr), function -[MGLRenderer getProgramBindingCount:type:], file MGLRenderer.m, line 1411.```

So I'll look at that next.

**Comment by darkaegisagain** [comment]
That is true, but you need something bound to effect a clear in Metal… I guess we could just detect this special case and bind an empty VAO.

Mike

**Comment by dokipen3d** [comment]
I can return a count of 0 from getProgramBindingCount if the vao is NULL (need to think if that is rigorous enough) and that lets the program continue, but no window is appearing. Onto the next hurdle.

**Comment by darkaegisagain** [comment]
If it gets to the metal layer you should be able to construct a render target etc and just do a clear.. but that would be fancy.

**Comment by darkaegisagain** [comment]
That would be fancy... I just bailed.

**Comment by r58Playz** [comment]
> This may or may not be a simple one. It seems that if we run the contents of test_clear function in a while loop, then we get this runtime error repeatedly.
> 
> `[CAMetalLayer nextDrawable] returning nil because allocation failed`
> 
> Not sure why, as other test programs can clear. They have extra stuff added though, so the process of adding programs and changing other state, must initialize metal correctly, whereas not doing those things makes clearing by itself not possible.
> 
> I'll keep digging and learn a few things long the way :)

Works fine now. It could have been caused by the IOSurface memory leak. According to apple docs about CAMetalLayer:
> If you don’t release drawables correctly, the layer runs out of drawables, and future calls to [nextDrawable](https://developer.apple.com/documentation/quartzcore/cametallayer/1478172-nextdrawable?language=objc) return nil.

**Comment by darkaegisagain** [comment]
It didn't seem to have issues when the window wasn't occluded... hmm... I thought drawables were reference counted like all other metal objects?
